File: app/src/main/java/com/jil/paintf/viewmodel/DocViewModel.java
package com.jil.paintf.viewmodel;

import androidx.lifecycle.MutableLiveData;
import com.jil.paintf.repository.*;
import com.jil.paintf.service.DataRoomService;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

import java.util.ArrayList;
import java.util.List;

public class DocViewModel extends BaseViewModel {
    RetrofitRepository retrofitRepository =RetrofitRepository.getInstance();
    private MutableLiveData<DocData> data;

    //==================================评论相关数据
    public MutableLiveData<ReplyRepository> liveReplyData;
    int pn=1;
    int maxpn=2;

    public DocViewModel() {
        liveReplyData =new MutableLiveData<>();
        data=new MutableLiveData<>();
    }

    public MutableLiveData<DocData> getData(){
        return data;
    }

    public void doNetGetDoc(int id){
        retrofitRepository.getDocDetail(id).subscribe(new Observer<DocRepository>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onNext(DocRepository docRepository) {
                data.postValue(docRepository.getData());
                saveHis(docRepository.getData());
            }

            @Override
            public void onError(Throwable e) {
                e.printStackTrace();
            }

            @Override
            public void onComplete() {

            }
        });
    }

    public void saveHis(final DocData docData){
        Observable<Integer> insert = Observable.create(new ObservableOnSubscribe<Integer>() {
            @Override
            public void subscribe(ObservableEmitter<Integer> emitter){
                int docid =docData.getItem().getDoc_id();
                String title =docData.getItem().getTitle();
                String image =docData.getItem().getPictures().get(0).getImg_src();
                int count =docData.getItem().getPictures().size();
                DataRoomService.getDatabase().getHisItemDao().insert(new HisItem(docid, image,title,count));
            }
        });

        add(insert.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(new Consumer<Integer>() {
            @Override
            public void accept(Integer integer) throws Exception {

            }
        }));

    }

    public void resetReply(){
        pn =1;
    }
    public void doNetReply(final int id) {
        if(pn!=-1)//-1 代表评论已经全部加载完成
        retrofitRepository.getDocReply(pn,id).subscribe(new Observer<ReplyRepository>() {
            @Override
            public void onSubscribe(Disposable d) {
            }

            @Override
            public void onNext(ReplyRepository replyRepository) {
                liveReplyData.postValue(replyRepository);
                if(replyRepository.getData().getPage().getCount()%20==0)
                    maxpn=replyRepository.getData().getPage().getCount()/20;
                else
                    maxpn=(replyRepository.getData().getPage().getCount()/20)+1;
            }

            @Override
            public void onError(Throwable e) {
                e.printStackTrace();
            }

            @Override
            public void onComplete() {
                if(pn<maxpn){
                    pn++;
                }else {
                    pn=-1;
                }
            }
        });
    }
}

File: app/src/main/java/com/jil/paintf/viewmodel/DocViewModel.kt
package com.jil.paintf.viewmodel

import android.graphics.drawable.Drawable
import androidx.lifecycle.MutableLiveData
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.jil.paintf.repository.*
import com.jil.paintf.service.AppPaintF
import com.jil.paintf.service.DataRoomService
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

class DocViewModel : BaseViewModel() {
    var retrofitRepository = RetrofitRepository.getInstance()
    val data: MutableLiveData<DocRepository>

    //==================================评论相关数据
    var liveReplyData: MutableLiveData<ReplyRepository> = MutableLiveData()
    var liveMyReply:MutableLiveData<Reply> = MutableLiveData()
    var pn = 1
    var maxpn = 2
    private fun doNetGetEmote() {
        retrofitRepository.emoteMap.subscribe { (_, data1) ->
            val packages = data1.packages
            for ((_, emotes) in packages) {
                for ((_, _, _, _, _, _, key, _, url) in emotes) {
                    Glide.with(AppPaintF.instance.applicationContext).asDrawable()
                        .load("$url@64w_64h.webp").into(object : CustomTarget<Drawable?>() {
                            override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable?>?) {
                                resource.setBounds(
                                    0,
                                    0,
                                    resource.intrinsicWidth,
                                    resource.intrinsicHeight
                                )
                                AppPaintF.instance.emoteMap[key] = resource
                            }

                            override fun onLoadCleared(placeholder: Drawable?) {}
                        })
                }
            }
        }.add()
    }

    fun doNetGetDoc(id: Int) {
        retrofitRepository.getDocDetail(id).subscribe(object : Observer<DocRepository> {
            override fun onSubscribe(d: Disposable) {}
            override fun onNext(docRepository: DocRepository) {
                data.postValue(docRepository)
                if(docRepository.code==0)
                    saveHis(docRepository.data)
            }

            override fun onError(e: Throwable) {
                e.printStackTrace()
            }

            override fun onComplete() {}
        })
    }

    fun saveHis(docData: DocData) {
        val insert =
            Observable.create<Int> {
                val docid = docData.item.doc_id
                val title = docData.item.title
                val image = docData.item.pictures[0].img_src
                val count = docData.item.pictures.size
                DataRoomService.getDatabase().hisItemDao.insert(HisItem(docid, image, title, count))
            }
        insert.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
            .subscribe { }.add()
    }

    fun resetReply() {
        pn = 1
    }

    fun doNetReply(id: Int) {
        if (pn != -1) //-1 代表评论已经全部加载完成
            retrofitRepository.getDocReply(pn, id).subscribe(object : Observer<ReplyRepository> {
                override fun onSubscribe(d: Disposable) {}
                override fun onNext(replyRepository: ReplyRepository) {
                    if(!replyRepository.data.replies.isNullOrEmpty()){
                        replyRepository.data.replies.map {
                            if(it.mid==AppPaintF.instance.loginId){
                                liveMyReply.postValue(it)
                            }
                        }
                    }

                    liveReplyData.postValue(replyRepository)
                    maxpn =
                        if (replyRepository.data.page.count % 20 == 0)
                            replyRepository.data.page.count / 20
                        else
                            replyRepository.data.page.count / 20 + 1
                }

                override fun onError(e: Throwable) {
                    e.printStackTrace()
                }

                override fun onComplete() {
                    if (pn < maxpn) {
                        pn++
                    } else {
                        pn = -1
                    }
                }
            })
    }

    init {
        data = MutableLiveData()
        if (AppPaintF.instance.emoteMap.size == 0) {
            //获取表情map
            doNetGetEmote()
        }
    }
}